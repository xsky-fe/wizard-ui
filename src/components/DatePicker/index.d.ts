import { DatePickerProps } from '../../interface';

export default function DatePicker(props: DatePickerProps): JSX.Element;