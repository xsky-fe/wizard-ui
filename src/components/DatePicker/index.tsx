import * as React from 'react';
import * as PropTypes from 'prop-types';
import Calendar from 'rc-calendar';
import RCDatePicker from 'rc-calendar/lib/Picker';
import TimePickerPanel from 'rc-time-picker/lib/Panel';
import 'rc-calendar/assets/index.css';
import 'rc-time-picker/assets/index.css';
import CALENDAR_ZH from 'rc-calendar/lib/locale/zh_CN';
import CALENDAR_EN from 'rc-calendar/lib/locale/en_US';
import moment, { Moment } from 'moment';
import { DatePickerProps } from '../../interface';

const format: string = 'YYYY-MM-DD HH:mm:ss';

function getFormat(time: boolean | undefined) {
  return time ? format : 'YYYY-MM-DD';
}

const DatePicker: React.FC<DatePickerProps> = props => {
  const {
    getTime,
    disabled,
    showTime,
    placeholder,
    reset,
    defaultValue,
    disabledDate,
    lang,
  } = props;
  const [val, setVal] = React.useState();
  const handleChange = React.useCallback(
    (value: Moment) => {
      setVal(value);
      if (getTime) {
        getTime(value && value.format(format));
      }
    },
    [getTime],
  );

  const timePickerElement = showTime ? <TimePickerPanel /> : null;
  const calendar = (
    <Calendar
      locale={lang === 'en' ? CALENDAR_EN : CALENDAR_ZH}
      timePicker={timePickerElement}
      disabledDate={disabledDate}
      showToday={false}
    />
  );

  return (
    <RCDatePicker
      animation="slide-up"
      disabled={disabled}
      calendar={calendar}
      value={reset ? null : val || (defaultValue && moment(defaultValue))}
      onChange={handleChange}
    >
      {({ value }: any) => {
        return (
          <span>
            <input
              placeholder={placeholder}
              disabled={disabled}
              readOnly
              value={(value && value.format(getFormat(showTime))) || ''}
            />
          </span>
        );
      }}
    </RCDatePicker>
  );
};

DatePicker.propTypes = {
  disabled: PropTypes.bool,
  showTime: PropTypes.bool,
  placeholder: PropTypes.string.isRequired,
  getTime: PropTypes.func.isRequired,
};

DatePicker.defaultProps = {
  disabled: false,
  showTime: true,
};

export default DatePicker;
