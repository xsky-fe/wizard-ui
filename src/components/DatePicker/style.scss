@import '../../style/variables.scss';
.rc-calendar-picker {
  z-index: 1060;
}
.DatePicker {
  input {
    height: 34px;
    padding: 6px 12px;
    font-size: $font-size-14;
    line-height: 1.42857;
    border: 1px solid $border-2;
    border-radius: 4px;
    box-shadow: inset 0 1px 1px rgba($black, 0.075);
    outline: none;
    &:focus {
      border-color: $primary-normal;
    }
    &:disabled {
      background-color: $bg-1;
      cursor: not-allowed;
    }
  }
}