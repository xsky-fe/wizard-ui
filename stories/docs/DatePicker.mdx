import { Canvas, Meta, ArgTypes } from '@storybook/blocks';

import * as DatePickerStories from '../DatePicker.stories.tsx';

<Meta of={DatePickerStories} />

# DatePicker
<em><strong>(wizard-ui自定义组件)</strong></em>
<p>2023-6-19 [peishan](https://github.xsky.com/peishan)</p>
展示月历，允许用户选择日期和具体的时间。基于 [rc-calendar](https://github.com/react-component/calendar) 和[rc-time-picker](https://github.com/react-component/time-picker)开发。

## 使用场景
1. 选择单个日期；
2. 默认支持选择日期和具体到秒级别的时间，可以定制只选择日期；
3. 支持中文、英文。
## 基本用法
### 默认展示
<Canvas of={DatePickerStories.Basic}/>
### 设置默认值
<Canvas of={DatePickerStories.DefaultValue}/>
### 设置 `lang` 为 `en` 展示为英文
<Canvas of={DatePickerStories.en}/>
### 禁用
<Canvas of={DatePickerStories.Disabled}/>
### 今天之后的禁止选择
<Canvas of={DatePickerStories.DisabledDate}/>
### 只选择日期
<Canvas of={DatePickerStories.NoTimeSelect}/>

<Canvas of={DatePickerStories.GetTime}/>

## API
<ArgTypes />