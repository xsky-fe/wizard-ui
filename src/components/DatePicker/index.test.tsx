import React from 'react';
import { mount } from 'enzyme';
import DatePicker from './index';
import renderer from 'react-test-renderer';

describe('DatePicker', () => {
  it('set defautValue', () => {
    const picker = mount(<DatePicker defaultValue="2019-08-13 12:02:53" />);
    const node = picker.find('input');
    expect(node.at(0).props().value).toBe('2019-08-13 12:02:53');
  });
  it('render show en lang', () => {
    const picker = renderer.create(<DatePicker lang="en" />).toJSON();
    expect(picker).toMatchSnapshot();
  });
  it('render disable', () => {
    const picker = renderer.create(<DatePicker disabled />).toJSON();
    expect(picker).toMatchSnapshot();
  });
  it('change time', () => {
    const picker = mount(<DatePicker />);
    const node = picker.find('input');
    node.simulate('click');
    picker
      .find('.rc-calendar-tbody > tr')
      .at(4)
      .find('td')
      .at(3)
      .simulate('click');
    expect(
      picker
        .find('.rc-calendar-input')
        .at(0)
        .props().value,
    ).not.toBe('');
  });
  it('get time', () => {
    const getTime = (val: string) => console.log(val);
    const picker = renderer.create(<DatePicker getTime={getTime} />).toJSON();
    expect(picker).toMatchSnapshot();
  })
});
