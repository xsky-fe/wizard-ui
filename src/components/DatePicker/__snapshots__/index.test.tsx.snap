// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatePicker get time 1`] = `
<span
  onClick={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  onTouchStart={[Function]}
>
  <input
    disabled={false}
    readOnly={true}
    value=""
  />
</span>
`;

exports[`DatePicker render disable 1`] = `
<span
  onKeyDown={[Function]}
>
  <input
    disabled={true}
    readOnly={true}
    value=""
  />
</span>
`;

exports[`DatePicker render show en lang 1`] = `
<span
  onClick={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  onTouchStart={[Function]}
>
  <input
    disabled={false}
    readOnly={true}
    value=""
  />
</span>
`;
